File: civitasRuby/lib/sorpresa_por_jugador.rb
# encoding:utf-8
module Civitas

  class SorpresaPorJugador < Sorpresa
  
    def initialize(texto, valor)
      super(texto)
      @valor = valor
    end

    def aplicar_a_jugador(actual, todos)
      if jugador_correcto(actual, todos)
        super(actual, todos)
        sorpresa_resta = SorpresaPagarCobrar.new("Pagan todos los jugadores 
menos el no afectado.", @valor * -1, )
        sorpresa_suma = SorpresaPagarCobrar("Recibe dinero de todos los 
jugadores.", @valor * (todos.size -1))
        for i in (0..todos.size)
          if i != actual
            sorpresa_resta.aplicar_a_jugador_pagar_cobrar(i, todos)
          else
            sorpresa_suma.aplicar_a_jugador_pagar_cobrar(i, todos)
          end
        end
      end
    end
    
    def to_string
      return super.to_string + "\nValor: #{@valor}"
    end
  
  public_class_method :new  
  end
end

File: civitas/src/civitas/SorpresaPorJugador.java
package civitas;


import java.util.ArrayList;

/**
 *
 * @author Usuario
 */
public class SorpresaPorJugador extends Sorpresa {
    
    private int valor;

    SorpresaPorJugador(int valor, String texto) {
        super(texto);
        this.valor = valor;
    }
    
    @Override
    public void aplicarAJugador(int actual, ArrayList<Jugador> todos) {
        if(jugadorCorrecto(actual, todos)) {
            informe(actual, todos);
            SorpresaPagarCobrar resta = new SorpresaPagarCobrar(valor*(-1), 
                    "Pagan todos los jugadores menos el no afectado.");
            for(int i = 0; i < todos.size(); i++) {
                if( i != actual) {
                    resta.aplicarAJugador(i, todos);
                }
            }
            SorpresaPagarCobrar suma = new SorpresaPagarCobrar(valor * 
                    (todos.size()-1), "Recibe dinero de todos los jugadores.");
            suma.aplicarAJugador(actual, todos);
        }
    }
    
    @Override
    public String toString() {
        String salida = "Valor: " + valor + super.toString();
        return salida;
    }
    
}